File: factors.py
#!/usr/bin/env python3

import sys

class Factorizer:
    """
    Class to factorize numbers into two smaller factors.
    """
    
    def __init__(self, filename):
        """
        Initialize the Factorizer instance with a filename.
        
        Args:
            filename (str): The name of the file containing numbers to factorize.
        """
        self.filename = filename
    
    def factorize_number(self, n):
        """
        Factorize a given number into two smaller factors.
        
        Args:
            n (int): The number to factorize.
        
        Returns:
            list: A list of tuples representing factor pairs.
        """
        factors = []
        for i in range(2, int(n**0.5) + 1):
            if n % i == 0:
                factors.append((i, n // i))
                break
        if not factors:
            factors.append((n, 1))
        return factors
    
    def factorize_file(self):
        """
        Factorize numbers from a file and print the factorization.
        """
        with open(self.filename, 'r') as file:
            for line in file:
                number = int(line.strip())
                factors = self.factorize_number(number)
                p, q = factors[0]
                print(f"{number}={p}*{q}")

if __name__ == '__main__':
    """
    Entry point of the program.
    """
    filename = sys.argv[1]
    factorizer = Factorizer(filename)
    factorizer.factorize_file()

File: factors.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <stdint.h>
#include <gmp.h>

// Pollard's rho algorithm for integer factorization
void factorize_number(const mpz_t n) {
    mpz_t x, y, d;
    mpz_inits(x, y, d, NULL);

    mpz_set_ui(x, 2);
    mpz_set_ui(y, 2);

    uint64_t i = 1;
    uint64_t k = 2;

    while (i < k) {
        mpz_mul(x, x, x);
        mpz_add_ui(x, x, 1);
        mpz_mod(x, x, n);

        mpz_mul(y, y, y);
        mpz_add_ui(y, y, 1);
        mpz_mod(y, y, n);

        mpz_sub(d, x, y);
        mpz_abs(d, d);

        mpz_gcd(d, d, n);

        if (mpz_cmp_ui(d, 1) != 0 && mpz_cmp(d, n) != 0) {
            mpz_t q;
            mpz_init(q);

            mpz_fdiv_q(q, n, d);

            gmp_printf("%Zd=%Zd*%Zd\n", n, d, q);

            mpz_clear(q);
            break;
        }

        if (i == k) {
            mpz_set(y, x);
            k <<= 1;
        }

        i++;
    }

    mpz_clears(x, y, d, NULL);
}

// Factorize numbers from a file
void factorize_file(const char* filename) {
    FILE* file = fopen(filename, "r");
    if (file == NULL) {
        printf("Failed to open file.\n");
        return;
    }

    char* line = NULL;
    size_t len = 0;

    while (getline(&line, &len, file) != -1) {
        mpz_t number;
        mpz_init(number);

        mpz_set_str(number, line, 10);

        factorize_number(number);

        mpz_clear(number);
    }

    free(line);
    fclose(file);
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        printf("Usage: ./factors <filename>\n");
        return 1;
    }

    const char* filename = argv[1];
    factorize_file(filename);

    return 0;
}


